src/elliptic.jl

```diff
@@ -0,0 +1,245 @@
+    if VERSION < v"0.7-"
+        sinu = sin(u)
+        cosu = cos(u)
+    else
+        sinu, cosu = sincos(u)
+    end
+    sn = sinu - m*(u-sinu*cosu)*cosu/4
+    cn = cosu + m*(u-sinu*cosu)*sinu/4
+    dn = 1 - m*sinu^2/2;
+    return sn,cn,dn
+end
+function ellipj_largem(u,m1)
+    # [1, Sec 16.15] / https://dlmf.nist.gov/22.10.ii
+    sinhu = sinh(u)
+    coshu = cosh(u)
+    tanhu = sinhu/coshu
+    sechu = 1/coshu
+    sn = tanhu + m1*(sinhu*coshu-u)*sechu^2/4
+    cn = sechu - m1*(sinhu*coshu-u)*tanhu*sechu/4
+    dn = sechu + m1*(sinhu*coshu+u)*tanhu*sechu/4
+    return sn,cn,dn
+end
+
+function ellipj_growm(sn,cn,dn, k)
+    # [1, Sec 16.12] / https://dlmf.nist.gov/22.7.i
+    for kk in reverse(k)
+        sn,cn,dn = (1+kk)*sn/(1+kk*sn^2),
+                       cn*dn/(1+kk*sn^2),
+                 (1-kk*sn^2)/(1+kk*sn^2)
+                 # ^ Use [1, 16.9.1]. Idea taken from [2]
+    end
+    return sn,cn,dn
+end
+function ellipj_shrinkm(sn,cn,dn, k::NTuple{N,<:Any}) where {N}
+    # [1, Sec 16.14] / https://dlmf.nist.gov/22.7.ii
+    for kk in reverse(k)
+        sn,cn,dn = (1+kk)*sn*cn/dn,
+                 (cn^2-kk*sn^2)/dn, # Use [1, 16.9.1]
+                 (cn^2+kk*sn^2)/dn  # Use [1, 16.9.1]
+    end
+    return sn,cn,dn
+end
+
+function ellipj_viasmallm(u,m,::Val{N}) where {N}
+    k,f,m = shrinkm(m,Val{N}())
+    sn,cn,dn = ellipj_smallm(u/f,m)
+    return ellipj_growm(sn,cn,dn,k)
+end
+function ellipj_vialargem(u,m,::Val{N}) where {N}
+    k,f,m1 = shrinkm(1-m,Val{N}())
+    sn,cn,dn = ellipj_largem(u/f,m1)
+    return ellipj_shrinkm(sn,cn,dn,k)
+end
+
+
+#----------------
+# Pick algorithm
+
+function ellipj_dispatch(u,m, ::Val{N}) where {N}
+    if abs(m) <= 1 && real(m) <= 0.5
+        return ellipj_viasmallm(u,m, Val{N}())
+    elseif abs(1-m) <= 1
+        return ellipj_vialargem(u,m, Val{N}())
+    elseif imag(m) == 0 && real(m) < 0
+        # [1, Sec 16.10]
+        sn,cn,dn = ellipj_dispatch(u*sqrt(1-m),-m/(1-m), Val{N}())
+        return sn/(dn*sqrt(1-m)), cn/dn, 1/dn
+    else
+        # [1, Sec 16.11]
+        sn,cn,dn = ellipj_dispatch(u*sqrt(m),1/m, Val{N}())
+        return sn/sqrt(m), dn, cn
+    end
+end
+
+Base.@pure puresqrt(x::Float64) = sqrt(x)
+Base.@pure function nsteps(m,ε)
+    i = 0
+    while abs(m) > ε
+        m = descstep(m)^2
+        i += 1
+    end
+    return i
+end
+Base.@pure nsteps(ε,::Type{<:Real}) = nsteps(0.5,ε) # Guarantees convergence in [-1,0.5]
+Base.@pure nsteps(ε,::Type{<:Complex}) = nsteps(0.5+sqrt(3)/2im,ε) # This is heuristic.
+function ellipj_nsteps(u,m)
+    # Compute the number of Landen steps required to reach machine precision.
+    # For all FloatXX types, this can be done at compile time, while for
+    # BigFloat this has to be done at runtime.
+    T = promote_type(typeof(u),typeof(m))
+    ε = puresqrt(Float64(eps(real(typeof(m)))))
+    N = nsteps(ε,typeof(m))
+    return ellipj_dispatch(u,m,Val{N}())::NTuple{3,T}
+end
```

Comment:
Hi. Sorry if this isn't in the correct place. I was hoping to this would be finished. From my reading of this pull request the major stumbling block in this pull request are these lines that use a lot of `@pure`. When I tried to test this I found that the `@pure` where needed to prevent dynamic dispatch. I tried to modify the code a bit to prevent this and wound up at the following

```suggestion
Base.@pure function nsteps(M)
    m = _convfac(M)
    ε = _vareps(M)
    i = 0
    while abs(m) > ε
        m = descstep(m)^2
        i += 1
    end
    return i
end

@inline _convfac(::Type{<:Real}) = 0.5
@inline _convfac(::Type{<:Complex}) = 0.5 + sqrt(3)/2im

@inline _vareps(::Type{<:Complex{T}}) where {T} = Float64(sqrt(eps(T)))
@inline _vareps(T::Type{<:Real}) = Float64(sqrt(eps(T)))

function ellipj_nsteps(u,m)
    # Compute the number of Landen steps required to reach machine precision.
    # For all FloatXX types, this can be done at compile time, while for
    # BigFloat this has to be done at runtime.
    T = promote_type(typeof(u),typeof(m))
    N = nsteps(T)
    return ellipj_dispatch(u,m,Val{N}())::NTuple{3,T}
end
```
This seems to work in my testing. For versions < 1.7 I still needed the `@pure` to prevent dynamic dispatch. I am not sure if this usage of `@pure` is alright, but it seems to be more inline with what I read in the docs. The good news is that for 1.7 everything infers correctly so the `@pure` is not needed. 

What's the chances that this gets implemented soonish? I can try to take over this pull request @ettersi  if any additional work is needed. I just need this implementation of the jacobi functions for some research I am doing.